---
title: "R Notebook"
output: html_notebook
editor_options: 
  chunk_output_type: console
  
params:
  h5: "" 
  ref: ""
  outFile: ""
---

```{r, handle_pkg, message=FALSE, echo=FALSE, include=FALSE}
# set library dir, load this and remove any other dirs to avoid confusion
# between personally created pkgs and the pipeline package
testing="Y"
if (testing=="Y"){
  # saving old path "/Library/Frameworks/R.framework/Versions/4.2-arm64/Resources/library"
  library_db_loc = "~/../../Volumes/CCBR_Pipeliner/db/PipeDB/scrna4.2Rlibs"
}else {
  library_db_loc = "/data/CCBR_Pipeliner/db/PipeDB/scrna4.2Rlibs"
}
assign(".lib.loc", library_db_loc, envir = environment(.libPaths))
.libPaths()

# set list of packages
list.of.packages=c("htmltools","Seurat","stringr","DoubletFinder","SingleR","scRNAseq",
                   "SingleCellExperiment","Orthology.eg.db","org.Mm.eg.db","org.Hs.eg.db",
                   "flexmix","SeuratWrappers","djvdj","Matrix")

#install as needed
new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[,"Package"])]
new.packages
if(length(new.packages)) BiocManager::install(new.packages)

# load packages
invisible(lapply(list.of.packages, library, character.only = TRUE))

```

```{r, prep_args, }
# set up params
h5 = params$h5 
ref= params$ref
outFile= params$outFile

# read in h5 file
rnaCounts = Read10X_h5(h5)

if(class(rnaCounts) == "list") {rnaCounts = rnaCounts$'Gene Expression'}

so <- CreateSeuratObject(rnaCounts)


groupFile = read.delim("groups.tab",header=F,stringsAsFactors = F)

sample = groupFile$V3[groupFile$V1 == tail(strsplit(h5,"/")[[1]],3)[1] & groupFile$V4 == "gex"]

print(groupFile[groupFile$V3 == sample & groupFile$V4 == "vdj",])

if (nrow(groupFile[groupFile$V3 == sample & groupFile$V4 == "vdj",]) > 0 ) {

tcrSamples = groupFile$V1[groupFile$V3 == sample & groupFile$V4 == "vdj"]
tcrSamples = paste0("cellrangerOut/",tcrSamples,"/outs")

so = import_vdj(input = so, vdj_dir = tcrSamples,  filter_paired = FALSE  )

```

