---
title: "R Notebook"
output: html_notebook
editor_options: 
  chunk_output_type: console
params:
  ref: "hg19"
  sampleid: "WB_Lysis_1"
  h5: "~/../../Volumes/data/scRNA/gex/WB_Lysis_1/WB_Lysis_1/outs/filtered_feature_bc_matrix.h5" 
  shared_source: "CCBR_Pipeliner/db/PipeDB/Rlibrary_4.2.2_scRNA"
---

```{r, prep_args,  message=FALSE, echo=FALSE, include=FALSE}
# set up params
ref= params$ref
sampleid = params$sampleid
h5 = params$h5 
shared_source=params$shared_source
```

```{r, handle_pkg, message=FALSE, echo=FALSE, include=FALSE}
# set library dir, load this and remove any other dirs to avoid confusion
# between personally created pkgs and the pipeline package
testing="Y"
if (testing=="Y"){
  # saving old path "/Library/Frameworks/R.framework/Versions/4.2-arm64/Resources/library"
  library_db_loc = paste0("~/../../Volumes/",shared_source)
}else {
  library_db_loc = paste0("/data/",shared_source)
}
assign(".lib.loc", library_db_loc, envir = environment(.libPaths))
.libPaths()

# set list of packages
list.of.packages=c("htmltools","stringr","DoubletFinder","SingleR","scRNAseq",
                   "SingleCellExperiment","Orthology.eg.db","org.Mm.eg.db","org.Hs.eg.db",
                   "flexmix","djvdj","Matrix")
# list.of.packages.bc=c("remotes","devtools")
list.of.packages.bc=c("hdf5r")
list.of.packages.cr=c("Seurat","R.utils")
list.of.packages.gh=c("satijalab/seurat-wrappers")
list.of.packages.gh_names=c("SeuratWrappers")

#install as needed
new.packages <- list.of.packages.bc[!(list.of.packages.bc %in% installed.packages()[,"Package"])]
new.packages
if(length(new.packages)) BiocManager::install(new.packages)

new.packages <- list.of.packages.cr[!(list.of.packages.cr %in% installed.packages()[,"Package"])]
new.packages
if(length(new.packages)) install.packages(new.packages)

new.packages <- list.of.packages.gh[!(list.of.packages.gh_names %in% installed.packages()[,"Package"])]
new.packages
if(length(new.packages)) remotes::install_github(new.packages)

# load packages
invisible(lapply(list.of.packages.bc, library, character.only = TRUE))
invisible(lapply(list.of.packages.cr, library, character.only = TRUE))
invisible(lapply(list.of.packages.gh_names, library, character.only = TRUE))

# source functions
source("scRNA_functions.R")
```

```{r, pre-processing, message=FALSE, echo=FALSE, include=FALSE, warning=FALSE}
# read in h5 file
rnaCounts = Read10X_h5(h5)

# if this is a multi-data type project, select only gex
if(class(rnaCounts) == "list") {
  print("--More than one data type received")
  rnaCounts = rnaCounts$'Gene Expression'
}

# crear object
so <- CreateSeuratObject(rnaCounts,min.cells=1)

# determine percent mito
so[["percent.mt"]] = PercentageFeatureSet(so, pattern = "^[M,m]T-")

# run QC, subset for features meeting critera
so <- RunMiQC(so, percent.mt = "percent.mt",
              nFeature_RNA = "nFeature_RNA", 
              posterior.cutoff = 0.7,
              model.slot = "flexmix_model") 
so_sub = subset(so, miQC.keep == "keep")
```

```{r, processing, message=FALSE, echo=FALSE, include=FALSE, warning=FALSE}
# Process data
so_processed = PROCESS_SO(so_sub)

# Label data
so_labeled = LABEL_SO(so_processed)

# ID doublets
dfso = doublets(so_labeled)

#label and filter
so_labeled$DF_hi.lo = dfso[[tail(names(dfso@meta.data),1)]]
so_singlet=subset(so_labeled,cells=names(so_labeled$DF_hi.lo)[so_labeled$DF_hi.lo =="Singlet"])

#save file
fpath=paste0(sampleid,".rds")
saveRDS(so_singlet,fpath)

```
