---
title: "scRNA Notebook"
output:
  pdf_document:
    toc: yes
editor_options:
  chunk_output_type: console
params:
  species: "hg19"
  sampleid: "WB_Lysis_1"
  h5: "cellranger_counts/sample1/outs/filtered_feature_bc_matrix.h5"
  qc_filtering: "manual"
  nCount_RNA_max: 500000
  nCount_RNA_min: 1000
  nFeature_RNA_max: 5000
  nFeature_RNA_min: 200
  percent_mt_max: 10
  percent_mt_min: 0
  run_doublet_finder: "N"
  npcs: 30
  scRNA_functions: "bin/scRNA_functions.R"
  testing: "N"
---

```{r setup}
options(rlang_trace_top_env = rlang::current_env())
options(error = function() {
  sink()
  print(rlang::trace_back(bottom = sys.frame(-1)), simplify = "none")
})
knitr::opts_chunk$set(message = FALSE)
# set up params
species <- params$species
sampleid <- params$sampleid
h5 <- params$h5

qc_filtering <- params$qc_filtering # manual, miqc, mads
nCount_RNA_max <- as.numeric(params$nCount_RNA_max)
nCount_RNA_min <- as.numeric(params$nCount_RNA_min)
nFeature_RNA_max <- as.numeric(params$nFeature_RNA_max)
nFeature_RNA_min <- as.numeric(params$nFeature_RNA_min)
percent_mt_max <- as.numeric(params$percent_mt_max)
percent_mt_min <- as.numeric(params$percent_mt_min)
run_doublet_finder <- params$run_doublet_finder # Y, N
npcs_val <- as.numeric(params$npcs)
testing <- params$testing

source(params$scRNA_functions)
library(AnnotationDbi)
library(celldex)
library(cluster)
library(DoubletFinder)
library(farver)
library(flexmix)
library(GenomeInfoDbData)
library(ggplot2)
library(hdf5r)
library(org.Hs.eg.db)
library(org.Mm.eg.db)
library(Orthology.eg.db)
library(Routliers)
library(RSQLite)
library(R.utils)
library(Seurat)
library(SeuratWrappers)
library(SingleCellExperiment)
library(SingleR)
library(tinytex)
```

```{r pre-processing}
# read in h5 file
rnaCounts <- Read10X_h5(h5)

# create S object
if (testing == "Y") {
  pbmc.data <- Read10X(data.dir = "/data/CCBR_Pipeliner/testdata/SINCLAIR/data_storage/10xFQ/filtered_gene_bc_matrices/hg19/")
  so <- CreateSeuratObject(counts = pbmc.data, project = "pbmc3k", min.cells = 3, min.features = 200)
} else {
  so <- CreateSeuratObject(counts = rnaCounts)
}
so

# add metadata
so$Sample <- sampleid

# determine percent mito
## http://htmlpreview.github.io/?https://github.com/satijalab/seurat-wrappers/blob/master/docs/miQC.html
so[["percent.mt"]] <- PercentageFeatureSet(so, pattern = "^mt-|^MT-")
so[["percent.RP"]] <- PercentageFeatureSet(so, pattern = "^R[PL,pl]|R[PS,ps]")
FeatureScatter(so, feature1 = "nFeature_RNA", feature2 = "percent.mt")

VlnPlot(so, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
```

```{r QC}
# subset
so_filt <- subset(so, subset = nFeature_RNA > 200)
so_filt

# run QC, subset for features meeting criteria
so_qc <- RunMiQC(so_filt,
  percent.mt = "percent.mt",
  nFeature_RNA = "nFeature_RNA",
  posterior.cutoff = 0.7,
  model.slot = "flexmix_model"
)
# PlotMiQC(so_qc, color.by = "miQC.probability") +
# ggplot2::scale_color_gradient(low = "grey", high = "purple")
```

```{r qc2}
# define feature info
nCount_out <- outliers_mad(so_qc$nCount_RNA, threshold = 3)$LL_CI_MAD
nFeature_out <- outliers_mad(so_qc$nFeature_RNA, threshold = 3)$LL_CI_MAD
mt_out <- outliers_mad(so_qc$percent.mt, threshold = 3)$UL_CI_MAD

# perform QC dependent on user choice
if (qc_filtering == "manual") {
  so_qc_select <- subset(
    so_qc,
    nCount_RNA_max >= nCount_RNA & nCount_RNA >= nCount_RNA_min &
      nFeature_RNA_max >= nFeature_RNA & nFeature_RNA >= nFeature_RNA_min &
      percent_mt_max >= percent.mt & percent.mt >= percent_mt_min
  )
} else if (qc_filtering == "miqc") {
  so_qc_select <- subset(so_qc, miQC.keep == "keep")
} else if (qc_filtering == "mads") {
  so_qc_select <- subset(so_qc,
    subset = nFeature_RNA > nFeature_out &
      nCount_RNA > nCount_out &
      percent.mt < mt_out
  )
}
so_qc_select

VlnPlot(so_qc_select, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
```

```{r process}
MAIN_PROCESS_SO <- function(so_in, species, npcs_in) {
  # assign genes depending on species input
  if (species == "hg38" || species == "hg19") {
    print("--proccesing human data")
    s.genes <- cc.genes$s.genes
    g2m.genes <- cc.genes$g2m.genes
  } else if (species == "mm10") {
    print("--proccesing mouse data")
    s.genes <- CONVERT_TO_HUMAN_GENELIST(cc.genes$s.genes)
    g2m.genes <- CONVERT_TO_HUMAN_GENELIST(cc.genes$g2m.genes)
  }

  # process
  so_1 <- NormalizeData(so_in,
    normalization.method = "LogNormalize",
    scale.factor = 10000,
    assay = "RNA"
  )
  so_2 <- ScaleData(so_1, assay = "RNA")
  so_3 <- CellCycleScoring(so_2,
    s.features = s.genes,
    g2m.features = g2m.genes,
    set.ident = TRUE
  )
  so_4 <- SCTransform(so_3)
  so_out <- SEURAT_CLUSTERING(so_4, npcs_in)
  return(so_out)
}

# Process data
so_processed <- MAIN_PROCESS_SO(so_qc_select, species, npcs_val)
```

```{r label}
MAIN_SINGLER <- function(so_in, species) {
  if (species == "hg38" || species == "hg19") {
    so_in$HPCA_main <- RUN_SINGLEr(so_in, celldex::HumanPrimaryCellAtlasData(), "label.main")
    so_in$HPCA <- RUN_SINGLEr(so_in, celldex::HumanPrimaryCellAtlasData(), "label.fine")
    so_in$BP_encode_main <- RUN_SINGLEr(so_in, celldex::BlueprintEncodeData(), "label.main")
    so_in$BP_encode <- RUN_SINGLEr(so_in, celldex::BlueprintEncodeData(), "label.fine")
    so_in$monaco_main <- RUN_SINGLEr(so_in, celldex::MonacoImmuneData(), "label.main")
    so_in$monaco <- RUN_SINGLEr(so_in, celldex::MonacoImmuneData(), "label.fine")
    so_in$immu_cell_exp_main <- RUN_SINGLEr(
      so_in, celldex::DatabaseImmuneCellExpressionData(),
      "label.main"
    )
    so_in$immu_cell_exp <- RUN_SINGLEr(
      so_in, celldex::DatabaseImmuneCellExpressionData(),
      "label.fine"
    )
    so_in$annot <- so_in$HPCA_main
  } else if (species == "mm10") {
    so_in$immgen_main <- RUN_SINGLEr(so_in, celldex::ImmGenData(), "label.main")
    so_in$immgen <- RUN_SINGLEr(so_in, celldex::ImmGenData(), "label.fine")
    so_in$mouseRNAseq_main <- RUN_SINGLEr(so_in, celldex::MouseRNAseqData(), "label.main")
    so_in$mouseRNAseq <- RUN_SINGLEr(so_in, celldex::MouseRNAseqData(), "label.fine")
    so_in$annot <- so_in$immgen_main
  }
  return(so_in)
}

# Label data
so_labeled <- MAIN_SINGLER(so_processed, species)
```
```{r doublets}
MAIN_DOUBLETS <- function(so_in, run_doublet_finder) {
  if (run_doublet_finder == "Y") {
    sweep.res.list_kidney <- paramSweep_v3(so_in, PCs = 1:10, sct = T)
    sweep.stats_kidney <- summarizeSweep(sweep.res.list_kidney, GT = FALSE)
    bcmvn_kidney <- find.pK(sweep.stats_kidney)

    ## Homotypic Doublet Proportion Estimate
    homotypic.prop <- modelHomotypic(so_in$annot)
    perc <- 0.005 * (length(colnames(so_in)) / 1000)
    nExp_poi <- round(perc * length(colnames(so_in))) # dfso@cell.names
    nExp_poi.adj <- round(nExp_poi * (1 - homotypic.prop))

    ## Run DoubletFinder with varying classification stringencies
    dfso <- doubletFinder_v3(so_in,
      pN = 0.25, pK = 0.09,
      nExp = nExp_poi,
      reuse.pANN = FALSE, PCs = 1:10, sct = T
    )

    pAAN <- tail(names(dfso@meta.data), 2)[1]
    dfso <- doubletFinder_v3(dfso,
      pN = 0.25, pK = 0.09,
      nExp = nExp_poi.adj,
      reuse.pANN = pAAN, PCs = 1:10, sct = T
    )
    so_in$DF_hi.lo <- dfso[[tail(names(dfso@meta.data), 1)]]
    so_in <- subset(so_in, cells = names(so_in$DF_hi.lo)[so_in$DF_hi.lo == "Singlet"])
  }

  return(so_in)
}

# ID doublets
so_doublet <- MAIN_DOUBLETS(so_labeled, run_doublet_finder)
```

```{r, save}
# clean final object
so_output <- UpdateSeuratObject(so_doublet)
so_output

VlnPlot(so_output, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3, group.by = "Sample")

# save file
fpath <- paste0(sampleid, "_seurat_preprocess.rds")
saveRDS(so_output, fpath)
```
