<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta name=description content="SINCLAIR is an open-source, reproducible solution for multiple single cell next-generation modalities"><meta name=author content="CCR Collaborative Bioinformatics Resource"><link href=../test-info/ rel=prev><link href=../troubleshooting/ rel=next><link rel=icon href=../../assets/images/favicon.png><meta name=generator content="mkdocs-1.6.1, mkdocs-material-9.1.6"><title>1. Differential Expression Analysis - SINCLAIR</title><link rel=stylesheet href=../../assets/stylesheets/main.ded33207.min.css><link rel=stylesheet href=../../assets/stylesheets/palette.a0c5b2b5.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script></head> <body dir=ltr data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo> <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#running-differential-expression class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <div data-md-color-scheme=default data-md-component=outdated hidden> </div> <header class=md-header data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=../.. title=SINCLAIR class="md-header__button md-logo" aria-label=SINCLAIR data-md-component=logo> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> SINCLAIR </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> 1. Differential Expression Analysis </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to dark mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_2 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6zm0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4zM7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"/></svg> </label> <input class=md-option data-md-color-media data-md-color-scheme=slate data-md-color-primary=red data-md-color-accent=red aria-label="Switch to light mode" type=radio name=__palette id=__palette_2> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3Z"/></svg> </label> </form> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </label> <nav class=md-search__options aria-label=Search> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg> </button> </nav> </form> <div class=md-search__output> <div class=md-search__scrollwrap data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list role=presentation></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/CCBR/SINCLAIR title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> CCBR/SINCLAIR </div> </a> </div> </nav> </header> <div class=md-container data-md-component=container> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class=md-grid> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=../.. class=md-tabs__link> Background </a> </li> <li class=md-tabs__item> <a href=../getting-started/ class=md-tabs__link> Usage </a> </li> <li class=md-tabs__item> <a href=./ class="md-tabs__link md-tabs__link--active"> Analysis </a> </li> <li class=md-tabs__item> <a href=../troubleshooting/ class=md-tabs__link> FAQ </a> </li> </ul> </div> </nav> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=../.. title=SINCLAIR class="md-nav__button md-logo" aria-label=SINCLAIR data-md-component=logo> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg> </a> SINCLAIR </label> <div class=md-nav__source> <a href=https://github.com/CCBR/SINCLAIR title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> CCBR/SINCLAIR </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../.. class=md-nav__link> Background </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2> <label class=md-nav__link for=__nav_2 id=__nav_2_label tabindex=0> Usage <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_2_label aria-expanded=false> <label class=md-nav__title for=__nav_2> <span class="md-nav__icon md-icon"></span> Usage </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../getting-started/ class=md-nav__link> 1. Getting Started </a> </li> <li class=md-nav__item> <a href=../preparing-files/ class=md-nav__link> 2. Preparing Files </a> </li> <li class=md-nav__item> <a href=../run/ class=md-nav__link> 3. Running the Pipeline </a> </li> <li class=md-nav__item> <a href=../output/ class=md-nav__link> 4. Expected Output </a> </li> <li class=md-nav__item> <a href=../test-info/ class=md-nav__link> 5. Running Test Data </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--active md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_3 checked> <label class=md-nav__link for=__nav_3 id=__nav_3_label tabindex=0> Analysis <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_3_label aria-expanded=true> <label class=md-nav__title for=__nav_3> <span class="md-nav__icon md-icon"></span> Analysis </label> <ul class=md-nav__list data-md-scrollfix> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> 1. Differential Expression Analysis <span class="md-nav__icon md-icon"></span> </label> <a href=./ class="md-nav__link md-nav__link--active"> 1. Differential Expression Analysis </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#choosing-the-identities class=md-nav__link> Choosing the identities </a> </li> <li class=md-nav__item> <a href=#preparing-the-seurat-object-for-differential-expression-with-prepsctfindmarkers class=md-nav__link> Preparing the Seurat object for differential expression with PrepSCTFindMarkers </a> </li> <li class=md-nav__item> <a href=#finding-differentially-expressed-genes-in-seurat class=md-nav__link> Finding differentially expressed genes in Seurat </a> <nav class=md-nav aria-label="Finding differentially expressed genes in Seurat"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#option-1-running-findmarkersref class=md-nav__link> Option 1: Running FindMarkersref </a> </li> <li class=md-nav__item> <a href=#option-2-running-findallmarkersref class=md-nav__link> Option 2: Running FindAllMarkersref </a> </li> <li class=md-nav__item> <a href=#alternative-methods-for-differential-expression class=md-nav__link> Alternative methods for differential expression </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#expected-outputs class=md-nav__link> Expected outputs </a> </li> <li class=md-nav__item> <a href=#common-issues-and-questions class=md-nav__link> Common issues and questions </a> <nav class=md-nav aria-label="Common issues and questions"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#why-are-there-so-many-genes-with-extremely-small-p-values class=md-nav__link> Why are there so many genes with extremely small p-values? </a> </li> <li class=md-nav__item> <a href=#requiring-joinlayers class=md-nav__link> Requiring JoinLayers </a> </li> <li class=md-nav__item> <a href=#running-differential-expression-on-subsets-of-cells class=md-nav__link> Running differential expression on subsets of cells </a> </li> <li class=md-nav__item> <a href=#pseudobulk-differential-expression class=md-nav__link> Pseudobulk differential expression </a> </li> </ul> </nav> </li> </ul> </nav> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4> <label class=md-nav__link for=__nav_4 id=__nav_4_label tabindex=0> FAQ <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_4_label aria-expanded=false> <label class=md-nav__title for=__nav_4> <span class="md-nav__icon md-icon"></span> FAQ </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../troubleshooting/ class=md-nav__link> Troubleshooting </a> </li> <li class=md-nav__item> <a href=../../contributing/ class=md-nav__link> How to contribute </a> </li> <li class=md-nav__item> <a href=../contributions/ class=md-nav__link> Contributors </a> </li> </ul> </nav> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <h1 id=running-differential-expression>Running Differential Expression<a class=headerlink href=#running-differential-expression title="Permanent link">&para;</a></h1> <p><em>Adapted from the <a href=https://satijalab.org/seurat/articles/de_vignette>Seurat differential expression vignette</a></em></p> <p>When running differential expression on the Seurat object produced by SINCLAIR, these steps should be followed to ensure an accurate output</p> <h2 id=choosing-the-identities>Choosing the identities<a class=headerlink href=#choosing-the-identities title="Permanent link">&para;</a></h2> <p>Before running any sort of differential expression, the identities of the Seurat object need to be determined and set by the user. For example, the following code will set the Seurat object identities to the clusters found at resolution 0.8:</p> <div class=highlight><pre><span></span><code>Idents(seuratObject) = &quot;SCT_snn_res.0.8&quot;
</code></pre></div> <p>The identities are not limited to the clusters; these can be set to any categorical variable in the metadata, including cell types that have been determined by SingleR or classified by the user, cell cycle phase (G1/G2M/S), or experimental group.</p> <h2 id=preparing-the-seurat-object-for-differential-expression-with-prepsctfindmarkers>Preparing the Seurat object for differential expression with <code>PrepSCTFindMarkers</code><a class=headerlink href=#preparing-the-seurat-object-for-differential-expression-with-prepsctfindmarkers title="Permanent link">&para;</a></h2> <p>After defining the object identities, Seurat requires that when when running differential expression on the SCT assay, the object needs to be prepared with the <code>PrepSCTFindMarkers</code> function:</p> <div class=highlight><pre><span></span><code>seuratObject = PrepSCTFindMarkers(seuratObject)
</code></pre></div> <p>The purpose of the <code>PrepSCTFindMarkers</code> is to use the minimum of the median UMI of individual objects to appropriately scale the SCT assay prior to differential expression<sup><a href=#refPrepSCTfindMarkers>ref</a></sup>.</p> <h2 id=finding-differentially-expressed-genes-in-seurat>Finding differentially expressed genes in Seurat<a class=headerlink href=#finding-differentially-expressed-genes-in-seurat title="Permanent link">&para;</a></h2> <p>Two options are available when running differential expression through Seurat.</p> <h3 id=option-1-running-findmarkersref>Option 1: Running <code>FindMarkers</code><sup><a href=#refFindMarkers>ref</a></sup><a class=headerlink href=#option-1-running-findmarkersref title="Permanent link">&para;</a></h3> <p>The "traditional" method of running differential expression compares two separate categories; this can still be exercised in the presence of more than the two comparison groups of interest. Following the example above, comparing the two largest clusters as determined by Seurat would be run as follows:</p> <div class=highlight><pre><span></span><code>deGeneList = FindMarkers(seuratObject,ident.1=0,ident.2=1,test.use = &quot;MAST&quot;)
</code></pre></div> <p>By default, Seurat uses the non-parametric Wilcoxon rank-sum test to identify significant genes. The MAST<sup><a href=#refMAST>ref</a></sup> algorithm uses a hurdle model to account for the sparsity of scRNASeq count matrices, and tends to be more sensitive to likely changes.</p> <h3 id=option-2-running-findallmarkersref>Option 2: Running <code>FindAllMarkers</code><sup><a href=#refFindAllMarkers>ref</a></sup><a class=headerlink href=#option-2-running-findallmarkersref title="Permanent link">&para;</a></h3> <p>When looking to define a set of potential clusters or identities using gene markers, the <code>FindAllMarkers</code> function will run the differential expression by setting the first identity group (equivalent to <code>ident.1</code>) to each cluster in turn and using the remaining clusters as the comparison group (<code>ident.2</code>).</p> <div class=highlight><pre><span></span><code>markerGeneList = FindAllMarkers(seuratObject,test.use=&quot;MAST&quot;)
</code></pre></div> <p>The resulting list will show the differentially expressed genes that are significantly enriched or depleted for each identity within the preselected category. In essence, <code>FindAllMarkers</code> runs a loop where <code>FindMarkers</code> is run for each individual identity.</p> <h3 id=alternative-methods-for-differential-expression>Alternative methods for differential expression<a class=headerlink href=#alternative-methods-for-differential-expression title="Permanent link">&para;</a></h3> <p>The default values for some of the more frequently altered parameters for the <code>FindMarkers</code> and <code>FindAllMarkers</code> functions are as follows:</p> <div class=highlight><pre><span></span><code>deGenes = FindMarkers(seuratObject, ident.1=NA, ident.2=NA, features=NULL,
  test.use=&quot;wilcox,&quot; logfc.threshold=0.25, min.pct=0.1)
</code></pre></div> <p>If <code>ident.1</code> is defined and <code>ident.2</code> is left as a NULL value, the cells not included in <code>ident.1</code> will be used as the comparison, i.e. essentially running a one vs. rest comparison, similar to what is described in <code>FindAllMarkers</code>.</p> <p>The <code>features</code> parameter can take a vector of specified genes and only run differential expression for those genes. Note that this assumes that the gene is not filtered out by any other criteria, such as those listed below. In all other cases, all genes are run through initial filtering and statistical testing.</p> <p>The <code>test.use</code> parameter is used to change the statistical test applied to identify differentially expressed genes. As mentioned above, the MAST algorithm is often applied to single cell data, while other tests that can be applied include <code>"bimod"</code>, <code>"roc"</code>, <code>"t"</code>, <code>"negbinom"</code>, <code>"poisson"</code>, <code>"LR"</code>, and <code>"DESeq2"</code>.</p> <p>The <code>logfc.threshold</code> filters out all genes that do not meet a log2 fold change threshold. This is applied to improve the speed of the computation, as those genes that do not meet this value are not subjected to statistical testing. The default threshold value is 0.25, which is roughly a 20% average fold change.</p> <p>The <code>min.pct</code> threshold filters out genes that are not expressed in the fraction of cells below the threshold for <strong>both</strong> comparison groups. If only one comparison group is below the threshold, the statistics for the gene will still be calculated. The default <code>min.pct</code> threshold is 0.1, or 10%.</p> <p>Since a number of genes will be unaffected by the experiment, a full differential expression list will require setting both <code>logfc.threshold</code> and <code>min.pct</code> to 0, so as to calculate the statistics for less relevant genes. These parameters may also need to be adjusted when selecting specific genes with the <code>features</code> parameter, as any user-defined genes that do not meet the threshold criteria will not be submitted to statistical testing.</p> <h2 id=expected-outputs>Expected outputs<a class=headerlink href=#expected-outputs title="Permanent link">&para;</a></h2> <p>The output of the <code>deGeneList</code> variable above will have a table structure resembling the following:</p> <table> <thead> <tr> <th></th> <th>p_val</th> <th>avg_log2FC</th> <th>pct.1</th> <th>pct.2</th> <th>p_val_adj</th> </tr> </thead> <tbody> <tr> <td>Tff1</td> <td>2.27649774378448e-20</td> <td>0.503646816090369</td> <td>0.313</td> <td>0.039</td> <td>5.15945448651315e-16</td> </tr> <tr> <td>Gkn1</td> <td>2.37538488060338e-18</td> <td>0.31537227402121</td> <td>0.23</td> <td>0.007</td> <td>5.3835722933995e-14</td> </tr> <tr> <td>Gkn2</td> <td>9.91735374467266e-14</td> <td>0.238583210040284</td> <td>0.166</td> <td>0.004</td> <td>2.24766905269261e-09</td> </tr> <tr> <td>Oaz1</td> <td>1.26240861948424e-12</td> <td>0.198987799763473</td> <td>0.996</td> <td>0.968</td> <td>2.86112289519909e-08</td> </tr> <tr> <td>Rab5c</td> <td>3.39748073842636e-12</td> <td>0.301705843310934</td> <td>0.645</td> <td>0.4</td> <td>7.7000503455695e-08</td> </tr> <tr> <td>Rbm3</td> <td>5.53653902770416e-12</td> <td>0.226144933696604</td> <td>0.97</td> <td>0.893</td> <td>1.25480120523887e-07</td> </tr> <tr> <td>Cfl1</td> <td>9.22315807122862e-12</td> <td>0.171800659275371</td> <td>1</td> <td>0.982</td> <td>2.09033654526325e-07</td> </tr> <tr> <td>Cd83</td> <td>2.27101797312635e-10</td> <td>0.242223105538207</td> <td>0.849</td> <td>0.789</td> <td>5.14703513429356e-06</td> </tr> <tr> <td>Tff2</td> <td>2.66531282660886e-10</td> <td>0.226607061165825</td> <td>0.189</td> <td>0.025</td> <td>6.04066499022633e-06</td> </tr> </tbody> </table> <p>The first unlabeled column lists the genes, followed by raw p-value, average log2 fold change, the percentage of cells expressing the gene in populations 1 and 2, and finally the adjusted p-value. The sign of the fold change and the definitions of <code>pct.1</code> and <code>pct.2</code> are defined by the order of groups selected in the <code>FindMarkers</code> call: Positive fold changes indicate enrichment in the first group (i.e. defined as <code>ident.1</code>), as well as the <code>pct.1</code> value</p> <h2 id=common-issues-and-questions>Common issues and questions<a class=headerlink href=#common-issues-and-questions title="Permanent link">&para;</a></h2> <h4 id=why-are-there-so-many-genes-with-extremely-small-p-values>Why are there so many genes with extremely small p-values?<a class=headerlink href=#why-are-there-so-many-genes-with-extremely-small-p-values title="Permanent link">&para;</a></h4> <p>In most circumstances, a small p-value generally indicates that the gene is extremely significant and should not be ignored. However, many of the statistical tests that are designed based on distributions, including MAST, operate under a set of assumptions. With single cell, one of the assumptions that gets violated is an upper limit on the number of replicates included, since each individual cell is treated as a replicate and leads to a comparison of 2 groups with as many as tens of thousands of replicates each.</p> <p>Unfortunately, it falls to the user to determine the relative significance of the genes that are identified as significant by examining the other statistics provided (i.e. <code>avg_log2FC</code> and <code>pct.1</code>/<code>pct.2</code>). Additionally, genes of interest can also be run through the <code>AverageExpression</code> and the <code>FeaturePlot</code> functions to explore the overall expression of the gene in the individual contrast groups. Other approaches, such as pseudobulk differential expression (see below) might be implemented in order to address this p-value phenomenon</p> <h4 id=requiring-joinlayers>Requiring <code>JoinLayers</code><a class=headerlink href=#requiring-joinlayers title="Permanent link">&para;</a></h4> <p>Since version 5, Seurat keeps individual samples as "layers" in the S4 data structure. This makes it simpler to apply various functions to each sample, such as SCTransform normalization or variable feature identification, since a single call to the Seurat object will behave like the <code>lapply</code> function in R. However, this also keeps the individual counts tables separate, which makes differential expression nigh impossible. To address this, the user needs to join the layers prior to running differential expression:</p> <div class=highlight><pre><span></span><code>so_dePrep = JoinLayers(seuratObject)
so_dePrep = PrepSCTFindMarkers(so_dePrep)
markers = FindAllMarkers(so_dePrep,test.use=&quot;MAST&quot;)
</code></pre></div> <h4 id=running-differential-expression-on-subsets-of-cells>Running differential expression on subsets of cells<a class=headerlink href=#running-differential-expression-on-subsets-of-cells title="Permanent link">&para;</a></h4> <p>Oftentimes the user will be interested in examining two different subsets of cells; for example, comparing the two experimental contrast groups in B cells alone. When isolating a group with the subset in question, the user might send a command such as:</p> <div class=highlight><pre><span></span><code>Bcells = subset(seuratObject,cells=colnames(seuratObject)[which(so$cellType==&quot;B cells&quot;)])
</code></pre></div> <p>However, when running <code>FindMarkers</code> on this subset, the following may be encountered:</p> <div class=highlight><pre><span></span><code>FindMarkers(Bcells,ident.1=&quot;group1&quot;,test.use=&quot;MAST&quot;)

Error in FindMarkers.SCTAssay(object = data.use, slot = slot, cells.1 = cells$cells.1,  :
  Object contains multiple models with unequal library sizes. Run `PrepSCTFindMarkers()` before running `FindMarkers()`.
</code></pre></div> <p>In this event, the user should ensure that the subset was extracted from the Seurat object that has already been prepared through <code>PrepSCTFindMarkers</code> to ensure appropriate scaling. The second step is to use a hidden parameter <code>recorrect_umi</code> to tell the program to "ignore" the scaling step.</p> <div class=highlight><pre><span></span><code>FindMarkers(Bcells,ident.1=&quot;group1&quot;,test.use=&quot;MAST&quot;, recorrect_umi=FALSE)
</code></pre></div> <p>In theory, this flag can also be used in all other steps in order to skip the scaling step for differential expression, but this is generally not recommended.</p> <h4 id=pseudobulk-differential-expression>Pseudobulk differential expression<a class=headerlink href=#pseudobulk-differential-expression title="Permanent link">&para;</a></h4> <p>If there are enough individual replicates, a user can convert the single cell dataset into a pseudobulk dataset and treat the individual samples as pooled replicates to be submitted through a standard RNASeq differential expression protocol, such as Limma or DESeq2. This has been explored using the <a href=https://github.com/neurorestore/Libra>Libra</a> tool and through <a href=https://satijalab.org/seurat/articles/de_vignette#perform-de-analysis-after-pseudobulking>Seurat</a> using their <code>AggregateExpression</code> function. The results tend to increase the p-values, which subsequently reduces the likelihood of false positives in identifying differentially expressed genes.</p> <p>The main warning when running a pseudobulk approach is to ensure that there are enough samples to warrant aggregation; if only one sample is available per experimental condition, the user will be attempting a 1v1 differential expression design, which is not nearly robust enough to account for any intra-group sample variability.</p> <p> </p> <p><font size=2>References:</p> <p><a name=refPrepSCTfindMarkers>1.</a> <a href=https://satijalab.org/seurat/reference/prepsctfindmarkers>PrepSCTFindMarkers</a></p> <p><a name=refFindMarkers>2.</a> <a href=https://satijalab.org/seurat/reference/findmarkers>FindMarkers</a></p> <p><a name=refFindAllMarkers>3.</a> <a href=https://satijalab.org/seurat/reference/findallmarkers>FindAllMarkers</a></p> <p></font></p> <p><font size=2>Author: Nathan Wong. January 2023</font></p> <hr> <div class=md-source-file> <small> Last update: 2025-02-26 </small> </div> </article> </div> </div> <button type=button class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg> Back to top </button> </main> <footer class=md-footer> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> <div class=md-copyright__highlight> Copyright &copy; 2023 CCR Collaborative Bioinformatics Resource </div> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-social> <a href=http://bioinformatics.cancer.gov target=_blank rel=noopener title=bioinformatics.cancer.gov class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 640 512"><!-- Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M144 0a80 80 0 1 1 0 160 80 80 0 1 1 0-160zm368 0a80 80 0 1 1 0 160 80 80 0 1 1 0-160zM0 298.7C0 239.8 47.8 192 106.7 192h42.7c15.9 0 31 3.5 44.6 9.7-1.3 7.2-1.9 14.7-1.9 22.3 0 38.2 16.8 72.5 43.3 96H21.3C9.6 320 0 310.4 0 298.7zM405.3 320h-.7c26.6-23.5 43.3-57.8 43.3-96 0-7.6-.7-15-1.9-22.3 13.6-6.3 28.7-9.7 44.6-9.7h42.7c58.9 0 106.7 47.8 106.7 106.7 0 11.8-9.6 21.3-21.3 21.3H405.3zM224 224a96 96 0 1 1 192 0 96 96 0 1 1-192 0zm-96 261.3c0-73.6 59.7-133.3 133.3-133.3h117.4c73.6 0 133.3 59.7 133.3 133.3 0 14.7-11.9 26.7-26.7 26.7H154.7c-14.7 0-26.7-11.9-26.7-26.7z"/></svg> </a> <a href=https://github.com/CCBR target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 496 512"><!-- Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg> </a> <a href=https://hub.docker.com/orgs/nciccbr/repositories target=_blank rel=noopener title=hub.docker.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 640 512"><!-- Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M349.9 236.3h-66.1v-59.4h66.1v59.4zm0-204.3h-66.1v60.7h66.1V32zm78.2 144.8H362v59.4h66.1v-59.4zm-156.3-72.1h-66.1v60.1h66.1v-60.1zm78.1 0h-66.1v60.1h66.1v-60.1zm276.8 100c-14.4-9.7-47.6-13.2-73.1-8.4-3.3-24-16.7-44.9-41.1-63.7l-14-9.3-9.3 14c-18.4 27.8-23.4 73.6-3.7 103.8-8.7 4.7-25.8 11.1-48.4 10.7H2.4c-8.7 50.8 5.8 116.8 44 162.1 37.1 43.9 92.7 66.2 165.4 66.2 157.4 0 273.9-72.5 328.4-204.2 21.4.4 67.6.1 91.3-45.2 1.5-2.5 6.6-13.2 8.5-17.1l-13.3-8.9zm-511.1-27.9h-66v59.4h66.1v-59.4zm78.1 0h-66.1v59.4h66.1v-59.4zm78.1 0h-66.1v59.4h66.1v-59.4zm-78.1-72.1h-66.1v60.1h66.1v-60.1z"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": "../..", "features": ["navigation.tabs", "navigation.top", "navigation.indexes", "toc.integrate"], "search": "../../assets/javascripts/workers/search.208ed371.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": {"provider": "mike"}}</script> <script src=../../assets/javascripts/bundle.51198bba.min.js></script> </body> </html>