<!DOCTYPE html><html class=writer-html5 lang=en> <head><meta charset=utf-8><meta http-equiv=X-UA-Compatible content="IE=edge"><meta name=viewport content="width=device-width, initial-scale=1.0"><meta name=author content="CCR Collaborative Bioinformatics Resource"><link rel="shortcut icon" href=../../img/favicon.ico><title>2. Preparing Files - SINCLAIR</title><link rel=stylesheet href=../../css/theme.css><link rel=stylesheet href=../../css/theme_extra.css><link rel=stylesheet href=https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css><link href=../../css/custom.css rel=stylesheet><link href=../../css/version-select.css rel=stylesheet><script>
        // Current page data
        var mkdocs_page_name = "2. Preparing Files";
        var mkdocs_page_input_path = "user-guide/preparing-files.md";
        var mkdocs_page_url = null;
      </script><!--[if lt IE 9]>
      <script src="../../js/html5shiv.min.js"></script>
    <![endif]--><script src=https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js></script><script>hljs.highlightAll();</script></head> <body class=wy-body-for-nav role=document> <div class=wy-grid-for-nav> <nav data-toggle=wy-nav-shift class="wy-nav-side stickynav"> <div class=wy-side-scroll> <div class=wy-side-nav-search> <a href=../.. class="icon icon-home"> SINCLAIR </a><div role=search> <form id=rtd-search-form class=wy-form action=../../search.html method=get> <input type=text name=q placeholder="Search docs" aria-label="Search docs" title="Type search term here"> </form> </div> </div> <div class="wy-menu wy-menu-vertical" data-spy=affix role=navigation aria-label="Navigation menu"> <ul> <li class=toctree-l1><a class="reference internal" href=../..>Background</a> </li> </ul> <p class=caption><span class=caption-text>Usage</span></p> <ul class=current> <li class=toctree-l1><a href=../getting-started/ class="reference internal">1. Getting Started</a> </li> <li class="toctree-l1 current"><a class="reference internal current" href=#>2. Preparing Files</a> <ul class=current> <li class=toctree-l2><a class="reference internal" href=#21-configs>2.1 Configs</a> <ul> <li class=toctree-l3><a class="reference internal" href=#211-nextflow-config>2.1.1 NextFlow Config</a> </li> <li class=toctree-l3><a class="reference internal" href=#212-base-config>2.1.2 Base Config</a> </li> <li class=toctree-l3><a class="reference internal" href=#213-modules-config>2.1.3 Modules Config</a> </li> <li class=toctree-l3><a class="reference internal" href=#214-r-package-config>2.1.4 R Package Config</a> </li> <li class=toctree-l3><a class="reference internal" href=#213-process-parameters>2.1.3 Process Parameters</a> </li> </ul> </li> <li class=toctree-l2><a class="reference internal" href=#22-preparing-manifests>2.2 Preparing Manifests</a> <ul> <li class=toctree-l3><a class="reference internal" href=#221-input-manifest>2.2.1 Input Manifest</a> </li> <li class=toctree-l3><a class="reference internal" href=#222-contrast-manifest>2.2.2 Contrast Manifest</a> </li> </ul> </li> </ul> </li> <li class=toctree-l1><a href=../run/ class="reference internal">3. Running the Pipeline</a> </li> <li class=toctree-l1><a href=../output/ class="reference internal">4. Expected Output</a> </li> <li class=toctree-l1><a href=../test-info/ class="reference internal">5. Running Test Data</a> </li> </ul> <p class=caption><span class=caption-text>Analysis</span></p> <ul> <li class=toctree-l1><a href=../differentialExpression/ class="reference internal">1. Differential Expression Analysis</a> </li> </ul> <p class=caption><span class=caption-text>FAQ</span></p> <ul> <li class=toctree-l1><a href=../troubleshooting/ class="reference internal">Troubleshooting</a> </li> <li class=toctree-l1><a href=../../contributing/ class="reference internal">How to contribute</a> </li> <li class=toctree-l1><a href=../contributions/ class="reference internal">Contributors</a> </li> </ul> </div> </div> </nav> <section data-toggle=wy-nav-shift class=wy-nav-content-wrap> <nav class=wy-nav-top role=navigation aria-label="Mobile navigation menu"> <i data-toggle=wy-nav-top class="fa fa-bars"></i> <a href=../..>SINCLAIR</a> </nav> <div class=wy-nav-content> <div class=rst-content><div role=navigation aria-label="breadcrumbs navigation"> <ul class=wy-breadcrumbs> <li><a href=../.. class="icon icon-home" aria-label=Docs></a></li> <li class=breadcrumb-item>Usage</li> <li class="breadcrumb-item active">2. Preparing Files</li> <li class=wy-breadcrumbs-aside> <a href=https://github.com/CCBR/SINCLAIR/edit/feature/gex_only/docs/user-guide/preparing-files.md>Edit on CCBR/SINCLAIR</a> </li> </ul> <hr> </div> <div role=main class=document itemscope=itemscope itemtype=http://schema.org/Article> <div class=section itemprop=articleBody> <h1 id=2-preparing-files>2. Preparing Files<a class=headerlink href=#2-preparing-files title="Permanent link">&para;</a></h1> <p>The pipeline is controlled through editing configuration and manifest files. Defaults are found in the /PIPELINEDIR/conf and /PIPELINEDIR/ directories</p> <p><img alt="SINCLAIR Process Overview" src="https://https://github.com/CCBR/SINCLAIR/blob/dev/assets/img/scRNA.jpeg?raw=true"> <sup><strong>Overview of Single Cell RNASeq Gene Expression Process</strong></sup></p> <h2 id=21-configs>2.1 Configs<a class=headerlink href=#21-configs title="Permanent link">&para;</a></h2> <p>The configuration files control parameters and software of the pipeline. These files are listed below:</p> <ul> <li>nextflow.config</li> <li>conf/base.config</li> <li>conf/modules.config</li> <li>conf/process_params.config</li> <li>conf/Rpack.config</li> </ul> <h3 id=211-nextflow-config>2.1.1 NextFlow Config<a class=headerlink href=#211-nextflow-config title="Permanent link">&para;</a></h3> <p>The configuration file dictates the global information to be used during the pipeline. Users can alter the default values, as needed: - input: path to input manifest; example manifests with (<code>input_manifest_cellranger.csv</code>) and without (<code>input_manifest.csv</code>) <code>cellranger</code> are included in assets - contrast: path to contrast manifest; example manifest (<code>contrast_manifest.csv</code>) is included in assets - outdir: path to output dir - species: species [options: hg19, mm10] - run_cellranger: determines whether to run cell ranger; if "Y" is selected, expects FQ inputs, if "N", expects .h5 inputs [options: "Y", "N"] - vars_to_regress: a comma separated list of any variables to regress during <code>SCTransform</code> process; [options: "", "percent.mt,nFeature_RNA,S.Score,G2M.Score,nCount_RNA"]</p> <h3 id=212-base-config>2.1.2 Base Config<a class=headerlink href=#212-base-config title="Permanent link">&para;</a></h3> <p>The configuration file dictates submission to Biowulf HPC. There are two different ways to control these parameters - first, to control the default settings, and second, to create or edit individual rules. These parameters should be edited with caution, after significant testing.</p> <h3 id=213-modules-config>2.1.3 Modules Config<a class=headerlink href=#213-modules-config title="Permanent link">&para;</a></h3> <p>The configuration file dictates process-specific processing parameters, including:</p> <ul> <li>the version of each software or program that is being used in the pipeline</li> <li>output location and file names</li> <li>additional arguments to be passed to the process</li> </ul> <h3 id=214-r-package-config>2.1.4 R Package Config<a class=headerlink href=#214-r-package-config title="Permanent link">&para;</a></h3> <p>The configuration file dictates which R libraries, and which versions, are loaded into the accompanying R script</p> <h3 id=213-process-parameters>2.1.3 Process Parameters<a class=headerlink href=#213-process-parameters title="Permanent link">&para;</a></h3> <p>The configuration file dictates process-specific user parameters, which varies for each process. Users can choose varied resolution values or QC methods, for example.</p> <h2 id=22-preparing-manifests>2.2 Preparing Manifests<a class=headerlink href=#22-preparing-manifests title="Permanent link">&para;</a></h2> <p>There are two manifests, which are required. These files describe information on the samples and desired contrasts. These files are:</p> <ul> <li>/assets/input_manifest.csv</li> <li>/assets/contrast_manifest.csv</li> </ul> <h3 id=221-input-manifest>2.2.1 Input Manifest<a class=headerlink href=#221-input-manifest title="Permanent link">&para;</a></h3> <p>This manifest will include information to sample level information. It includes the following column headers:</p> <ul> <li>masterID: This is the biological sample ID; duplicates are allowed in this column</li> <li>uniqueID: This is a unique sample level ID; duplicates are not allowed in this column</li> <li>groupID: This is the groupID which should match to the <code>contrast_manifest</code>; duplicates are allowed in this column</li> <li>dataType: This is the datatype for the input sample; options are 'gex' 'atac' 'vdj'</li> <li>input_dir: This is the input directory for the data files of the sample type (IE "/path/to/sample1/fastq")</li> </ul> <p>An example sampleManifest file is shown below:</p> <p>| masterID | uniqueID | groupID | dataType | input_dir | | ---------- | -------- | ------- | --------------------------------------------------------------------- | ----------------------------------------------------------------- | ------------------------------------------------------- | | WB_Lysis_1 | sample1 | group1 | gex | /data/CCBR_Pipeliner/Pipelines/TechDev_scRNASeq_Dev2023/test_dir/ | WB_Lysis_Granulocytes_3p_Introns_8kCells_fastqs/sample1 | | WB_Lysis_1 | sample2 | group1 | gex | /data/CCBR_Pipeliner/Pipelines/TechDev_scRNASeq_Dev2023/test_dir/ | WB_Lysis_Granulocytes_3p_Introns_8kCells_fastqs/sample2 | | WB_Lysis_2 | sample3 | group2 | gex | /data/CCBR_Pipeliner/Pipelines/TechDev_scRNASeq_Dev2023/test_dir/ | WB_Lysis_Granulocytes_3p_Introns_8kCells_fastqs/sample3 | | WB_Lysis_2 | sample4 | group2 | gex | /data/CCBR_Pipeliner/Pipelines/TechDev_scRNASeq_Dev2023/test_dir/ | WB_Lysis_Granulocytes_3p_Introns_8kCells_fastqs/sample4 | | WB_Lysis_3 | sample5 | group3 | gex,/data/CCBR_Pipeliner/Pipelines/TechDev_scRNASeq_Dev2023/test_dir/ | WB_Lysis_Granulocytes_3p_Introns_8kCells_fastqs/sample5 | | WB_Lysis_1 | sample6 | group1 | atac | /data/CCBR_Pipeliner/Pipelines/TechDev_scRNASeq_Dev2023/test_dir/ | WB_Lysis_Granulocytes_3p_Introns_8kCells_fastqs/sample1 |</p> <h3 id=222-contrast-manifest>2.2.2 Contrast Manifest<a class=headerlink href=#222-contrast-manifest title="Permanent link">&para;</a></h3> <p>This manifest will include sample information to performed differential comparisons. A few requirements:</p> <ul> <li>groups listed must match groups within the <code>input_manifest</code> groupID column</li> <li>headers should be included for the max number of contrasts. In the example below, the second contrast contains 3 groups, and so the header includes contrast1-contrast3</li> <li>multiple groups can be added by increasing the header and adding additional contrasts, as needed</li> </ul> <p>An example contrast file:</p> <table> <thead> <tr> <th>contrast1</th> <th>contrast2</th> <th>contrast3</th> </tr> </thead> <tbody> <tr> <td>group1</td> <td>group2</td> <td></td> </tr> <tr> <td>group1</td> <td>group2</td> <td>group3</td> </tr> </tbody> </table> </div> </div><script async src="https://www.googletagmanager.com/gtag/js?id=G-D3SL9V30KL"></script> <script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());
  gtag('config', 'G-D3SL9V30KL'); // Replace with your tracking ID
</script> </div> </div> </section> </div> <div class=rst-versions role=note aria-label=Versions> <span class=rst-current-version data-toggle=rst-current-version> <span> <a href=https://github.com/CCBR/SINCLAIR class="fa fa-code-fork" style="color: #fcfcfc"> CCBR/SINCLAIR</a> </span> <span><a href=../getting-started/ style="color: #fcfcfc">&laquo; Previous</a></span> <span><a href=../run/ style="color: #fcfcfc">Next &raquo;</a></span> </span> </div> <script src=../../js/jquery-3.6.0.min.js></script> <script>var base_url = "../..";</script> <script src=../../js/theme_extra.js></script> <script src=../../js/theme.js></script> <script src=../../js/custom.js></script> <script src=../../search/main.js></script> <script src=../../js/version-select.js></script> <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script> </body> </html> 